docs: deployment flow architecture (#1470)
keptn/lifecycle-toolkit: docs/content/en/docs/concepts/architecture/deployment-flow.md

---
title: Flow of deployment
linktitle: Flow of deployment
description: Understand the execution flow of a deployment
weight: 35
---

The Keptn Lifecycle Toolkit (KLT) deploys a
[Kubernetes Workload](https://kubernetes.io/docs/concepts/workloads/)
by passing through a well-defined execution flow.

The execution flow goes through six main phases:

* Pre-deployments-tasks
* Pre-deployment-evaluation
* Deployment
* Post-deployment-tasks
* Post-deployment-evaluation
* Completed

> Within each phase,

Reviewer: Actually, the tasks within each phase are all executed in parallel

Author: done

the order of execution is determined
by the order in which evaluations and tasks are listed in the
[KeptnApp](../../yaml-crd-ref/app/)
resource.

## Kubernetes events

> A [Kubernetes Event](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/)
> is emitted at each phase

Reviewer: There can potentially be multiple events that are sent during a phase, so it does not necessarily always have to be one:

```suggestion
[Kubernetes Events](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/)
are emitted at each phase
```

to provide additional Observability of the execution flow.

The Keptn Lifecycle Toolkit implements a
[Permit Scheduler Plugin](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#permit)
that blocks the binding of the pods to a node
until all the pre-conditions are fulfilled.

> A Kubernetes deployment is started by the following command:

Reviewer: ```suggestion
A Kubernetes deployment is applied to the cluster by the following command:
```

```shell
kubectl apply -f deployment.yaml
```

> TODO: Should this mention anything

Author: Text modified appropriately.

AppDeploy
  AppDeployStarted
  WorkloadPreDeployTasks
    WorkloadPreDeployTasksStarted
    WorkloadPreDeployTasksSucceeded OR WorkloadPreDeployTasksErrored 
  WorkloadPreDeployEvaluations
    WorkloadPreDeployEvaluationsStarted
    WorkloadPreDeployEvaluationsSucceeded OR WorkloadPreDeployErrored
  WorkloadDeploy
    WorkloadDeployStarted
    WorkloadDeploySucceeded OR WorkloadDeployErrored
  WorkloadPostDeployTasks
    WorkloadPostDeployTasksStarted
    WorkloadPostDeployTasksSucceeded OR WorkloadPostDeployTasksErrored 
  WorkloadPostDeployEvaluations
    WorkloadPostDeployEvaluationsStarted
    WorkloadPostDeployEvaluationsSucceeded OR WorkloadPostDeployEvaluationsErrored
  AppDeploySucceeded OR AppDeployErrored
  ```

### Post-deployment phase

TODO: Add intro statement about this phase.

```shell
AppPostDeployTasks
  AppPostDeployTasksStarted
  AppPostDeployTasksSucceeded OR AppPostDeployTasksErrored
```

### Post-deployment evaluation phase

TODO: Add intro statement about this phase.

```shell
AppPostDeployEvaluations
  AppPostDeployEvaluationsStarted
  AppPostDeployEvaluationsSucceeded OR AppPostDeployEvaluationsErrored
```

### Completed phase

```shell
Completed
```

> ## Events that are not part of the deployment flow

Author: odubajDT 

I might think that this section is a little bit misleading. The list below represents phases and during these phases, there might be some events send, but it's not a given, for example for Complete. Maybe a better description would be something the the sense that we have such phases and in some of them out-of-order events describing the state/changes in the system are triggered

Additional phases/states exist,
such as those that describe what happens when something fails.

Whenever something in the system happens (we create a new resource, etc.)
a Kubernetes event is generated.
The following events are defined as part of the Keptn Lifecycle Toolkit
but they are not part of the deployment flow.
These include:

```shell
CreateEvaluation
ReconcileEvaluation
ReconcileTask
CreateTask
CreateApp
CreateAppVersion
CreateWorkload
CreateWorkloadInstance
Completed
Deprecated
WorkloadDeployReconcile

> WorkloadDeployReconcileErrored

Author: odubajDT commented:

Actually I was not able to find these two phases nor events in the system

---
title: Flow of deployment
linktitle: Flow of deployment
description: Understand the execution flow of a deployment
weight: 35
---

The Keptn Lifecycle Toolkit (KLT) deploys a
[Kubernetes Workload](https://kubernetes.io/docs/concepts/workloads/)
by passing through a well-defined execution flow.

The execution flow goes through six main phases:

* Pre-deployments-tasks
* Pre-deployment-evaluation
* Deployment
* Post-deployment-tasks
* Post-deployment-evaluation
* Completed

Within each phase,
all tasks are executed in parallel.
They are not affected by the order
in which evaluations and tasks are listed in the
[KeptnApp](../../yaml-crd-ref/app.md/)

> resource.

Reviewer 2: or in the annotations `keptn.sh/pre-tasks` and `keptn.sh/pre-evaluations` (same for post items`

Author: done

## Kubernetes events

> [Kubernetes Events](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/)

Reviewer 2: ```suggestion
[Kubernetes Events](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/) and [CloudEvents](https://cloudevents.io/)
```

are emitted at each phase
to provide additional Observability of the execution flow.

The Keptn Lifecycle Toolkit implements a
[Permit Scheduler Plugin](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#permit)
that blocks the binding of the pods to a node
until all the pre-conditions are fulfilled.

A Kubernetes deployment is started by the deployment engine
that is implemented
(such as Flux or Argo)
or can be started by the following command:

```shell
kubectl apply -f deployment.yaml
```

KLT does not care how a deployment manifest is applied to the cluster.
Both `kubectl` and Flux/Argo send the manifest to the Kubernetes API
so KLT does not differentiate the actual deployment options.
This also means that one Keptn Application
can include services that are deployed with different methods.

The deployment is created
but the created pods are blocked and in pending state
until all the required pre-deployment tasks/evaluations
defined on either the `KeptnApp` or `KeptnWorkload` level pass.
Only then are the pods bound to a node and deployed.
If any evaluation or task fails,
the `KeptnApp` issues an appropriate `*Errored` event

> and terminates the deployment.

Reviewer 3: ```suggestion
and the deployment will stay indefinetly pending until further changes or external intervention
```

If all evaluations and tasks in a phase are successful,
the `KeptnApp` issues the appropriate `*Succeeded` event
and initiates the next phase.

## Summary of deployment flow

To view these events on your cluster, execute:

```shell
kubectl get events -n <namespace> . 
```

### Pre-deployment phase

> TODO: Add intro statement about this phase.

Reviewer 3: ```suggestion
Pre-deployment tasks can perform any kind of action needed to prepare for the deployment, but also tests, load tests or similar.
```

```shell
AppPreDeployTasks
  AppPreDeployTasksStarted
  AppPreDeployTasksSucceeded OR AppPreDeployTasksErrored
```

### Pre-deployment evaluation phase

> TODO: Add intro statement about this phase.

Reviewer 3: ```suggestion
Pre deployment evaluation can be used to assert the status of the cluster or of services the workload depends on, to assure it will be deployed only if some prerequisites are met. 
```

The `KeptnApp` just observes whether
all pre and post-deployment tasks/evaluation are successful
and that the pods are deployed successfully.
When all activities are successful,
the `KeptnApp` issues the `AppDeploySucceeded` event
and continues to the next phase.
If any of these activities fail,
the `KeptnApp` issues the `AppDeployErrored` event
and terminates the deployment.

```shell
AppDeploy
  AppDeployStarted
  WorkloadPreDeployTasks
    WorkloadPreDeployTasksStarted
    WorkloadPreDeployTasksSucceeded OR WorkloadPreDeployTasksErrored 
  WorkloadPreDeployEvaluations
    WorkloadPreDeployEvaluationsStarted
    WorkloadPreDeployEvaluationsSucceeded OR WorkloadPreDeployErrored
  WorkloadDeploy
    WorkloadDeployStarted
    WorkloadDeploySucceeded OR WorkloadDeployErrored
  WorkloadPostDeployTasks
    WorkloadPostDeployTasksStarted
    WorkloadPostDeployTasksSucceeded OR WorkloadPostDeployTasksErrored 
  WorkloadPostDeployEvaluations
    WorkloadPostDeployEvaluationsStarted
    WorkloadPostDeployEvaluationsSucceeded OR WorkloadPostDeployEvaluationsErrored
  AppDeploySucceeded OR AppDeployErrored
  ```

### Post-deployment phase

> TODO: Add intro statement about this phase.

Reviewer 3: ```suggestion
Post deployment is the tipical phase where one can run test on the freshly deployed application, for instance to gather performance data.
```

```shell
AppPostDeployTasks
  AppPostDeployTasksStarted
  AppPostDeployTasksSucceeded OR AppPostDeployTasksErrored
```

### Post-deployment evaluation phase

> TODO: Add intro statement about this phase.

Reviewer 3: ```suggestion
The post-deployment evaluation can be used to analyze the cluster/application status after the new workload deployment. The result of this phase does not revert or influence the deployment but can be used by other external tools, for instance, to react on a failure.
```

The deployment is created
but the created pods are blocked and in pending state
until all the required pre-deployment tasks/evaluations
defined on either the `KeptnApp` or `KeptnWorkload` level pass.
Only then are the pods bound to a node and deployed.
If any pre-deployment evaluation or task fails,
the `KeptnApp` issues an appropriate `*Errored` event
and the deployment remains pending indefinitely,
until further changes or external intervention
If all evaluations and tasks in a phase are successful,
the `KeptnApp` issues the appropriate `*Succeeded` event
and initiates the next phase.

## Summary of deployment flow

To view these events on your cluster, execute:

```shell
kubectl get events -n <namespace> . 
```

### Pre-deployment phase

Pre-deployment tasks can perform any kind of action needed
to prepare for the deployment, including unit tests, load tests or other similar tests.

```shell
AppPreDeployTasks
  AppPreDeployTasksStarted
  AppPreDeployTasksSucceeded OR AppPreDeployTasksErrored
```

### Pre-deployment evaluation phase

Pre-deployment evaluation can be used to assert the status of the cluster
or of services the workload depends on,

> to assure it is deployed only if the specified prerequisites are met. 

Author: ```suggestion
to assure it is deployed only if the specified prerequisites are met.
```
